app/master/cluster_master.py

```diff
@@ -55,6 +62,22 @@ def __init__(self):
 
         SlavesCollector.register_slaves_metrics_collector(lambda: self.all_slaves_by_id().values())
 
+    def configure_heartbeat(self):
+        self._heartbeat_job = self._scheduler.add_job(self.heartbeat, 'interval', seconds=self._heartbeat_frequency)
+        return self._scheduler
+
+    def heartbeat(self):
+        t = datetime.datetime.now()
+        slaves_to_disconnect = []
+        for slave in self._all_slaves_by_url.values():
+            if slave.is_alive() and not slave.is_responsive(t, self._heartbeat_frequency):
+                slaves_to_disconnect.append(slave)
+
+        for slave in slaves_to_disconnect:
+            self._disconnect_slave(slave)
+            self._logger.warning('Slave {} marked offline as it is not sending heartbeats.'.format(
```

Comment:
This should probably be more than a warning. I think this should be logged as an error and/or increase a Prometheus error metric.